lib/avromatic.rb

```diff
@@ -77,19 +77,38 @@ def self.prepare!(skip_clear: false)
         end
       end
     end
-    eager_load_models!
+    eager_load_models!(skip_nested_models: skip_clear)
   end
 
   def self.eager_load_models=(models)
     @eager_load_model_names = Array(models).map { |model| model.is_a?(Class) ? model.name : model }
   end
 
-  def self.eager_load_models!
+  def self.eager_load_models!(skip_nested_models: true)
     (@eager_load_model_names || []).each do |model_name|
-      nested_models.ensure_registered_model(model_name.constantize)
+      model = model_name.constantize
+      next if !nested_models.ensure_registered_model(model) || skip_nested_models
+
+      model.attribute_definitions.each do |_attribute, definition|
+        eager_load_nested_models!(definition.type)
+      end
     end
   end
   private_class_method :eager_load_models!
+
+  def self.eager_load_nested_models!(type)
+    case type
+    when Avromatic::Model::Types::ArrayType, Avromatic::Model::Types::MapType
+      eager_load_nested_models!(type.value_type)
+    when Avromatic::Model::Types::CustomType
+      eager_load_nested_models!(type.default_type) # correct?
+    when Avromatic::Model::Types::UnionType
+      type.member_types.each { |member_type| eager_load_nested_models!(member_type) }
+    when Avromatic::Model::Types::RecordType
+      nested_models.ensure_registered_model(type.record_class)
```

Comment:
Begs the question if we should restart the recursion with the `#attribute_definitions` of `type.record_class`.